### ISSUE TITLE
DHCP - Setting "Gateway" to "None" On a Specific Device Does Not Work

### ISSUE THREAD

**Issue by NateroniPizza** [opnsense/core#6343]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [X ] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [X ] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

I see that this was previously brought up a couple years back (https://github.com/opnsense/core/issues/4843), but that was when this was initially added (because it was not available). At this point, the feature looks to no longer be working.

**Describe the bug**

I am attempting to set a static DHCP mapping for one of my clients without a gateway (this is on a dual-NIC device, and to ensure anything outside of the NICs' local subnets get routed out a specific NIC, I want the secondary NIC to not receive a gateway when pulling DHCP). When I set the "gateway" of that device's static mapping to "none," as specified in the "gateway" description, it still gives it a gateway. If I set this same setting to "none" on the DHCP interface context, it is correctly applied (i.e. no gateway assigned when pulling DHCP on client).

**The current OPNsense version where the bug first appeared:** Currently running OPNsense 23.1_6-amd64
**The last OPNsense version where the bug did not exist:** Don't know. I've only been running it since the previous version, and it wasn't working there too.
**The exact URL of the GUI page involved (if any):** https://10.0.20.1/services_dhcp_edit.php?if=opt3&id=0 (installation-specific. To get there go to Services>DHCPv4>[select interface]>Edit existing DHCP client).

*Tip*: to validate your setup was working with the previous version, use opnsense-revert (https://docs.opnsense.org/manual/opnsense_tools.html#opnsense-revert)

**To Reproduce**

1. Service>DHCPv4>[your interface]
2. Set up new static DHCP mapping.
3. Set "Gateway" for the static mapping to "none," per the Gateway description.
4. Save and renew IP on DHCP client.

**Expected behavior**

No gateway assigned to DHCP client.

**Describe alternatives you considered**

None available, aside from setting static IP.

**Screenshots**

![image](https://user-images.githubusercontent.com/14255967/220013829-9dc0d4e5-50b2-46d8-abdf-d7b52becc1b2.png)

**Relevant log files**

n/a

**Additional context**

n/a

**Environment**

Software version used and hardware type if relevant, e.g.:

Client: Windows 11
OPNSense version: OPNsense 23.1_6-amd64

**Comment by fichtner** [comment]
Hmm, I'm not convinced about the "requirement" here and the "fix". The none was introduced in https://github.com/opnsense/core/issues/4843 but it just mimics what empty selection does... DHCPd will likely still inherit from the pool setting so you can't unset a previously set gateway anyway...

**Comment by fichtner** [comment]
PS: It would be nice to have a link to the forum post in the ticket as well...

**Comment by fichtner** [comment]
> If that's all their "fix" did, then they missed the point of that previous GitHub ticket.

I really don’t agree with this.

**Comment by NateroniPizza** [comment]
> > If that's all their "fix" did, then they missed the point of that previous GitHub ticket.
> 
> I really don’t agree with this.

"A DHCPv4 service on an interface allows setting 'gateway=none' globally. When creating a static DHCP assignment for a single host/ip/mac, under gateway overrides for that host 'gateway=none' throws an error.

This is seen as unexpected, I should be able to override this one host's gateway setting that may be global. "none" works when set at interface level - therefore an override should be possible at host level."

Quote from previous ticket. They say the functionality of "gateway=none" works as expected "globally" (in the Interface context) - for anyone that reads the "Gateway" description, them tests it, it is clear that this does exactly what it says - it does not assign a "gateway" to DHCP clients. They then say that this does not work on a per-device context. I don't know how much more clearly they could have stated it. For someone to then just change the description to say it will do exactly what is done at an interface context, then allow it to just ignore the setting, they obviously missed the point.

Though I don't understand the code well enough to know whether that is what the change did. For all I know, they could have fixed it, then it been broken on a subsequent update.

**Comment by fichtner** [comment]
Honestly, just let me know what configuration in dhcpd.conf you want vs. what you currently get. I suspect there is a problem in the lease info given out to the client containing a router and you want to get rid of it, but I am unsure you understand what you are asking. I'll quote for emphasis:

> DHCPd will likely still inherit from the pool setting so you can't unset a previously set gateway

**Comment by NateroniPizza** [comment]
> Honestly, just let me know what configuration in dhcpd.conf you want vs. what you currently get. I suspect there is a problem in the lease info given out to the client containing a router and you want to get rid of it, but I am unsure you understand what you are asking. I'll quote for emphasis:
> 
> > DHCPd will likely still inherit from the pool setting so you can't unset a previously set gateway

Let me try explaining it another way...

Just to be clear, in case I wasn't, when I or the previous ticket are saying "none," we aren't meaning that the field should be left blank - we are referring to a string of "none" (without quotes) being entered into the "Gateway" address. This should result in the DHCP response being sent by OPNSense omitting the existence of the "Default Gateway." This is how it works at the DHCP global/interface context (under Services>DHCPv4>[LAN]). This is not how it is working under when, on that same page, you create a new or edit an existing DHCP static mapping - it is just ignoring an input of "none."

I (and the previous person that asked for this) want it to work on a per-device context the same way it works on the global/interface context.

**_Screenshots showing how it works when "Gateway" is configured on the DHCP settings at the global/interface context:_**
**Global=blank:**
![image](https://user-images.githubusercontent.com/14255967/220171265-652dc765-9936-4629-a845-8840cfdbda2b.png)

**Global="none"**
![image](https://user-images.githubusercontent.com/14255967/220172460-6a7acfe4-a814-4181-abe1-e76a30f83533.png)

**_Screenshots of how it is working when it is instead set at the device context within the specific device's DHCP Static Mapping context:_**
**Global=blank**
![image](https://user-images.githubusercontent.com/14255967/220171993-51dc7182-9b68-434c-9a10-d0c890c96f53.png)

**Global="none"**
![image](https://user-images.githubusercontent.com/14255967/220171999-548fb0af-9012-414b-b4a4-aab9f79d02e3.png)

**Comment by AdSchellevis** [comment]
you can find the generated dhcp config at `/var/dhcpd/etc/dhcpd.conf` on the firewall, their manual should help you around in figuring out the possibilities and impossibilities (https://kb.isc.org/docs/isc-dhcp-44-manual-pages-dhcpd)

**Comment by NateroniPizza** [comment]
Thanks for pointing me to that file @AdSchellevis .

I believe I'm seeing what is meant by "DHCPd will likely still inherit from the pool setting so you can't unset a previously set gateway." I see that you can override the "option routers" value with another valid IP, but I'm not sure whether this is a null value or something you can use to override the pool's "option routers" value with a blank value.

Is there any way to manually edit the dhcpd.conf file and not have it get overwritten when the DHCP service is restarted? I'd like to try some things (like "option routers null;" "option routers ;" and find whatever a "null" character would be and put that after it, to see if any of these options are available).

**Comment by NateroniPizza** [comment]
It looks like this can be accomplished by editing the dhcp-parameter-request-list to specify which DHCP fields are sent back to a client (and, by extension, allowing you to omit the "router")... If there is a way to expose this option in the GUI, then that could work as well.

https://forums.freebsd.org/threads/dhcp-how-to-exclude-router-in-ip-address-reservation-in-isc-dhcpd.87840/

**Comment by AdSchellevis** [comment]
> Is there any way to manually edit the dhcpd.conf file and not have it get overwritten when the DHCP service is restarted?

Not very user-friendly, but you could kill the service manually and start it in a similar way as the script does:

https://github.com/opnsense/core/blob/08fb2ea495c3c54eb39d50550639a4cbec41e8e1/src/etc/inc/plugins.inc.d/dhcpd.inc#L1213

As mentioned in the other ticket (https://github.com/opnsense/core/issues/4843), "none" is only added for consistency, the effect is the same as empty.

https://github.com/opnsense/core/blob/08fb2ea495c3c54eb39d50550639a4cbec41e8e1/src/etc/inc/plugins.inc.d/dhcpd.inc#L1116-L1118

**Comment by fichtner** [comment]
Not sure why sending rockets to the moon is a good idea. Just add a pool where your static mapping IP is in range and then set gateway "none" for that pool and leave gateway in static mapping empty?

**Comment by NateroniPizza** [comment]
> Not sure why sending rockets to the moon is a good idea. Just add a pool where your static mapping IP is in range and then set gateway "none" for that pool and leave gateway in static mapping empty?

This could work, but I am running into the same problem as adding it on the per-device context - it is still globally defined, and thus inherited by the Additional Pool.

![image](https://user-images.githubusercontent.com/14255967/220184396-c9c19198-4e58-4b77-9907-ba70a0d225e6.png)

If I switch Gateway to "none" to omit it from the global context, then set an Additional Range for the usual DHCP range and configure the Gateway to 10.0.10.1 in there, it does work. However, any devices with static mappings outside of the pool will need the Gateway added to them individually. I am also getting a "[wifi SSID] has no internet access" showing up in a notification on my Android device when it is configured, despite being able to reach the internet on it.

![image](https://user-images.githubusercontent.com/14255967/220183903-43343724-16e5-498c-aef5-6ed1b68091f2.png)

> As mentioned in the other ticket (#4843), "none" is only added for consistency, the effect is the same as empty.
> 
> https://github.com/opnsense/core/blob/08fb2ea495c3c54eb39d50550639a4cbec41e8e1/src/etc/inc/plugins.inc.d/dhcpd.inc#L1116-L1118

Consistency of messages is not ideal if the message is misleading from an end user's perspective. Now that I understand what this is doing behind the scenes, I can understand why it would be thought of as consistency - however, from what an an end-user sees in the WebUI, it appears to be a bug. Entering "none" in both of these contexts, as far as the configuration goes, does similar - it simply omits the "option routers" parameter in these two contexts. However, they effectively do very different things: In the global pool context, it **blocks inheriting of the gateway address**, whereas in the per-device DHCP mapping context, it does nothing.

To an end user (i.e. me) reading the description, "Type "none" for no gateway assignment" seems to very clearly state that the client being configured will not receive a default gateway. Then when one reads that exact same description in the global/interface context, and sees it behave in the expected manner (no gateway address being sent to clients), then this appears to be a bug.

**Comment by fichtner** [comment]
> Then when one reads that exact same description in the global/interface context, and sees it behave in the expected manner (no gateway address being sent to clients), then this appears to be a bug.

I can fix the help test in a second, but it's not the solution you want. I really need you to stay on point which is how dhcpd.conf can be adjusted accordingly. Everything else is just demotivation or waste of time.


Cheers,
Franco

**Comment by NateroniPizza** [comment]
> > Then when one reads that exact same description in the global/interface context, and sees it behave in the expected manner (no gateway address being sent to clients), then this appears to be a bug.
> 
> I can fix the help text in a second, but it's not the solution you want. I really need you to stay on point which is how dhcpd.conf can be adjusted accordingly. Everything else is just demotivation or waste of time.
> 
> Cheers, Franco

Agreed - that won't fix the underlying behavior. In the end, if this is a limitation of dhcpd, the message needs to be addressed, though.



Ok, I think I've found a value that works in the dhcpd.conf. Thank you for referring me to the way to start the service @AdSchellevis . While the WebUI's value isn't being pulled from this config file, I see that the changes I make in the dhcpd.conf are being applied to clients.

Default:
![image](https://user-images.githubusercontent.com/14255967/220187242-cdf5da49-74f1-4597-a76d-797875034a67.png)

Adding "option routers null;" to the specific device:
![image](https://user-images.githubusercontent.com/14255967/220187292-d684d5e2-a5bf-44b9-8d0d-7a8b541aca71.png)


One potential concern with this, is that this is likely just assuming any non-IP values in there are hostnames. Since "null" isn't a hostname I have configured in DNS, it is getting back a blank IP, and so the DHCP value passed out for Default Gateway is blank. When I put a valid hostname in there, though, that is the IP that gets passed. I would assume that a device having a hostname of "null" would be bad practice anyway, so this could potentially be a way to handling "none" values to keep a default gateway blank.
